Adding add token signing certificate content to v1 (#14137)
microsoftgraph/microsoft-graph-docs: api-reference/v1.0/api/serviceprincipal-addtokensigningcertificate.md

---
title: "servicePrincipal: addTokenSigningCertificate"
description: "Add a signing certificate to a servicePrincipal."
ms.localizationpriority: medium
author: "davidmu"
ms.prod: "applications"
doc_type: "apiPageType"
---

# servicePrincipal: addTokenSigningCertificate

Namespace: microsoft.graph

> Creates a self-signed signing certificate and returns a [selfSignedCertificate](../resources/selfsignedcertificate.md), which is the public part of the generated certificate. The self-signed signing certificate is composed of these resources: the private key ([keyCredential](../resources/keycredential.md) with usage = 'Sign'), the public key ([keyCredential](../resources/keycredential.md) with usage = 'verify'), and the [passwordCredential](../resources/passwordcredential.md). All the created resources have the same **customKeyIdentifier**.

Reviewer: ```suggestion
Creates a self-signed signing certificate and returns a [selfSignedCertificate](../resources/selfsignedcertificate.md) object, which is the public part of the generated certificate. The self-signed signing certificate is composed of the following objects which are added to the [servicePrincipal](../resources/serviceprincipal.md): 
+ The [keyCredentials](../resources/keycredential.md) object with the following objects:
    + A private key object with **usage** set to `Sign`.
    + A public key object with **usage** set to `Verify`.
+ The [passwordCredentials](../resources/passwordcredential.md) object. 

All the objects have the same value of **customKeyIdentifier**.
```

> The **passwordCredential** is used to open the pfx/private key. Also, it's associated with the privateKey having the same **keyId**. The subject of the certificate is a constant value. It won't be affected by the optional **displayName** provided in the POST call. The **startDateTime** is set to the same time the certificate is created using the action. The **endDateTime** can be up to three years after the certificate is created.

Reviewer: ```suggestion
The **passwordCredential** is used to open the PFX file (private key). It and the associated private key object have the same value of **keyId**. Once set during creation through the **displayName** property, the subject of the certificate cannot be updated. The **startDateTime** is set to the same time the certificate is created using the action. The **endDateTime** can be up to three years after the certificate is created.
```

## Permissions

|Permission type      | Permissions (from least to most privileged)              |
|:--------------------|:---------------------------------------------------------|
|Delegated (work or school account) | Application.ReadWrite.All, Directory.ReadWrite.All |
|Delegated (personal Microsoft account) | None.    |
|Application | Application.ReadWrite.OwnedBy, Application.ReadWrite.All, Directory.ReadWrite.All |

## HTTP request

<!-- { "blockType": "ignored" } -->

```http
POST /servicePrincipals/{id}/addTokenSigningCertificate
```

## Request body

In the request body, provide the following required properties.

| Property	   | Type	|Description|
|:---------------|:--------|:----------|
| displayName | string | Friendly name for the key.  It must start with `CN=`.|

> | endDateTime | DateTimeOffset |The date and time when the credential expires. It can be up to 3 years from the date the certificate is created. If not supplied, the default is three years from the time of creation. The Timestamp type represents date and time information using ISO 8601 format and is always in UTC time. For example, midnight UTC on Jan 1, 2014 would look like this: '2014-01-01T00:00:00Z' .|

Reviewer: ```suggestion
| endDateTime | DateTimeOffset |The date and time when the credential expires. It can be up to 3 years from the date the certificate is created. If not supplied, the default is three years from the time of creation. The timestamp type represents date and time information using ISO 8601 format and is always in UTC time. For example, midnight UTC on Jan 1, 2014 is `2014-01-01T00:00:00Z`.|
```

## Response

If successful, this method returns a `200 OK` response code and a new [selfSignedCertificate](../resources/selfsignedcertificate.md) object in the response body.

## Examples

### Request

The following is an example of the request.

> # [HTTP](#tab/http)

Reviewer: Code snippets at auto-generated thanks to this statement `"blockType": "request"` in Line 57. Do not add this manually. See [the guidance](https://dev.azure.com/msazure/One/_wiki/wikis/Microsoft%20Graph%20Partners/103162/Generate-code-snippets) for more.
```suggestion
```

<!-- {
  "blockType": "request",
  "name": "serviceprincipal_addtokensigningcertificate"
}-->

```http

> POST https://graph.microsoft.com/beta/servicePrincipals/004375c5-6e2e-4dec-95e3-626838cb9f80/addTokenSigningCertificate

Reviewer: ```suggestion
POST https://graph.microsoft.com/v1.0/servicePrincipals/004375c5-6e2e-4dec-95e3-626838cb9f80/addTokenSigningCertificate
```

Content-type: application/json

{
    "displayName":"CN=customDisplayName",
    "endDateTime":"2024-01-25T00:00:00Z"
}
```

> # [C#](#tab/csharp)
> [!INCLUDE [sample-code](../includes/snippets/csharp/serviceprincipal-addtokensigningcertificate-csharp-snippets.md)]
> [!INCLUDE [sdk-documentation](../includes/snippets/snippets-sdk-documentation-link.md)]
> 
> # [JavaScript](#tab/javascript)
> [!INCLUDE [sample-code](../includes/snippets/javascript/serviceprincipal-addtokensigningcertificate-javascript-snippets.md)]
> [!INCLUDE [sdk-documentation](../includes/snippets/snippets-sdk-documentation-link.md)]
> 
> # [Objective-C](#tab/objc)
> [!INCLUDE [sample-code](../includes/snippets/objc/serviceprincipal-addtokensigningcertificate-objc-snippets.md)]
> [!INCLUDE [sdk-documentation](../includes/snippets/snippets-sdk-documentation-link.md)]
> 
> # [Java](#tab/java)
> [!INCLUDE [sample-code](../includes/snippets/java/serviceprincipal-addtokensigningcertificate-java-snippets.md)]
> [!INCLUDE [sdk-documentation](../includes/snippets/snippets-sdk-documentation-link.md)]
> 
> ---

Reviewer: Same comment about code snippets. Do not add these snippets & the closing `---` manually. See [the guidance](https://dev.azure.com/msazure/One/_wiki/wikis/Microsoft%20Graph%20Partners/103162/Generate-code-snippets) for more.
```suggestion
```

### Response

The following is an example of the response.

<!-- {
  "blockType": "response",
  "truncated": true,
  "@odata.type": "microsoft.graph.selfSignedCertificate"
} -->

```http
HTTP/1.1 200 OK
Content-Type: application/json

{

>   "@odata.context": "https://graph.microsoft.com/beta/$metadata#microsoft.graph.selfSignedCertificate",

Reviewer: ```suggestion
  "@odata.context": "https://graph.microsoft.com/v1.0/$metadata#microsoft.graph.selfSignedCertificate",
```
